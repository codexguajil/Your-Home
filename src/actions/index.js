export const hasErrored = (message) => ({
  type: 'HAS_ERRORED',
  message
});

export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
});

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});