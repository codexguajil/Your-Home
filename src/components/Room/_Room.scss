.room {
  width: 40%;
}

.room:nth-child(odd) {  
  margin-right: 20%;
}