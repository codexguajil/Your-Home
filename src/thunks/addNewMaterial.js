import { isLoading, hasErrored, addMaterial } from '../actions';

export const addNewMaterial = (body) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify(body),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      console.log(data)
      dispatch(isLoading(false))
      // dispatch(addMaterial(data.data.createProject))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}