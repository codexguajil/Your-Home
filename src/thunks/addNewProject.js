
import { isLoading, hasErrored, addProject } from '../actions';

export const addNewProject = (body) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify(body),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      dispatch(isLoading(false))
      dispatch(addProject(data.data.createProject))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}
