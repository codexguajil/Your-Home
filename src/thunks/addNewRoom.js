import { isLoading, hasErrored, addRoom } from '../actions';

export const addNewRoom = (id, name, description, type) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({"query": `mutation { addProjectRoom (project_id: "${id}", room: {name: "${name}", description: "${description}", type: "${type}"}) {id name description type project {id}}}`}),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json();
      console.log(data.data.addProjectRoom);
      dispatch(isLoading(false));
      dispatch(addRoom(data.data.addProjectRoom));
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}