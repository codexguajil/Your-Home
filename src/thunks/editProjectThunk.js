// { "query":"mutation{updateProject(project: {id: 2, name: \"Updated House\", city: \"Different City\"}) {id name city state}}"
// }

// name, address, description

import { isLoading, hasErrored, editProject } from '../actions';

export const editProjectThunk = (id, name, address, description) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({
          "query": `mutation {updateProject (project: {id: "${id}", name: "${name}", address: "${address}", description: "${description}"}) {id name address description}}`
          }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      console.log(data.data.updateProject)
      dispatch(isLoading(false))
      dispatch(editProject(data.data.updateProject))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}