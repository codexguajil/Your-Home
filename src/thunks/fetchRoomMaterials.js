import { isLoading, hasErrored, setMaterials } from '../actions';


export const fetchRoomMaterials = (id) => {
  return async (dispatch) => {
    const url = "https://hometrackr.herokuapp.com//api/v1/graphql"
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        body: JSON.stringify({
          "query": `query { getRoomsMaterials(room_id: ${id})}`
          }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        }
      })
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const data = await response.json()
      console.log(data.data)
      dispatch(isLoading(false))
      dispatch(setMaterials(data.data.getRoomsMaterials))
    } catch(error) {
      dispatch(hasErrored(error.message))
    }
  }
}