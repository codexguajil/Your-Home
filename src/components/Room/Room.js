import React from 'react';
import { connect } from 'react-redux';
import { Drawer, List, Button } from 'antd';
import Materials from '../../containers/Materials/Materials';
import EditRoom from '../../containers/EditRoom/EditRoom';
import { fetchRoomMaterials } from '../../thunks/fetchRoomMaterials';
import DeleteConfirm from '../DeleteConfirm/DeleteConfirm';

export class Room extends React.Component {
  state = { visible: false };

  showDrawer = async () => {
    const { id } = this.props
    await this.props.fetchRoomMaterials(id)
    this.setState({ visible: true });
  }

  onClose = () => {
    this.setState({ visible: false });
  }

  render() {
    const { name, type, description, materials } = this.props;
    return (
      <div className='room'>
        <List
          dataSource={[
            {
              name: name,
              description: description,
            }
          ]}
          bordered
          renderItem={item => (
            <List.Item key={item.id} className='room-item'actions={[
              <Button type='link' onClick={this.showDrawer}>View Materials</Button>,
              <EditRoom {...this.props}/>,
              <DeleteConfirm type="room" id={item.id} />
            ]}>
              <List.Item.Meta
                title={item.name}
                description={item.description}
              />
            </List.Item>
          )}
        />
        <Drawer
          width={640}
          placement="right"
          closable={false}
          onClose={this.onClose}
          visible={this.state.visible}
        >
          <Materials type={type}
                     materials={materials}
          />
        </Drawer>
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  fetchRoomMaterials: (materials) => dispatch(fetchRoomMaterials(materials))
});

export const mapStateToProps = state => ({
  materials: state.materials
});

export default connect(mapStateToProps, mapDispatchToProps)(Room);