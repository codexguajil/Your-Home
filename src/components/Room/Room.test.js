import React from 'react';
import { shallow } from 'enzyme';
import { Room } from './Room';
import { setMaterials } from '../../actions';
import { mapStateToProps, mapDispatchToProps } from './Room';
import { fetchRoomMaterials } from '../../thunks/fetchRoomMaterials';

jest.mock('../../thunks/fetchRoomMaterials')

describe('Room', () => {
  let wrapper;
  let mockFn;

  beforeEach(() => {
    mockFn = jest.fn();
    wrapper = shallow(<Room fetchRoomMaterials={mockFn}/>)
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should gather materials when showDrawer is called', () => {
    wrapper.instance().showDrawer();
    expect(mockFn).toHaveBeenCalled()
  });

  it('should setState on showDrawer', async () => {
    wrapper.setState({ visible: false })
    await wrapper.instance().showDrawer();
    expect(wrapper.state('visible')).toBe(true);
  });

  it('should setState on onClose', () => {
    wrapper.setState({ visible: true })
    wrapper.instance().onClose();
    expect(wrapper.state('visible')).toBe(false);
  });

  it('should return a state object', () => {
    const mockState = {
      projects: [],
      materials: {},
      randomState: {}
    }
    const expected = {
      materials: {}
    }

    const mockProps = mapStateToProps(mockState)
    expect(mockProps).toEqual(expected)
  });

  it('should mapDispatchToProps', () => {
    const mockMaterials = { shower: [], sink: []}
    const mockDispatch = jest.fn()
    const actionToDispatch = fetchRoomMaterials(mockMaterials)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.fetchRoomMaterials(mockMaterials)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  });

});