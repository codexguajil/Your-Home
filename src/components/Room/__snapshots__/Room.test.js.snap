// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room should match snapshot 1`] = `
<div
  className="room"
>
  <List
    bordered={true}
    dataSource={
      Array [
        Object {
          "description": undefined,
          "name": undefined,
        },
      ]
    }
    loading={false}
    pagination={false}
    renderItem={[Function]}
    split={true}
  />
  <withConfigConsumer(Drawer)
    closable={false}
    onClose={[Function]}
    placement="right"
    visible={false}
    width={640}
  >
    <RoomItems />
  </withConfigConsumer(Drawer)>
</div>
`;
